src/Middleware/RequestAutoPopulationMiddleware.php

```
@@ -0,0 +1,96 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+namespace Google\ApiCore\Middleware;
+
+use Google\ApiCore\Call;
+use GuzzleHttp\Promise\PromiseInterface;
+use Ramsey\Uuid\Uuid;
+
+/**
+ * Middleware that adds autopopulation functionality. This middlware is
+ * added iff auto population settings are present in the resource
+ * descriptor config for the rpc method in context.
+ *
+ * @internal
+ */
+class RequestAutoPopulationMiddleware implements MiddlewareInterface
+{
+    /** @var callable */
+    private $nextHandler;
+
+    /** @var array<string, string> */
+    private $autoPopulationSettings;
+
+    public function __construct(
+        callable $nextHandler,
+        array $autoPopulationSettings
+    ) {
+        $this->nextHandler = $nextHandler;
+        $this->autoPopulationSettings = $autoPopulationSettings;
+    }
+
+    /**
+     * @param Call $call
+     * @param array $options
+     *
+     * @return PromiseInterface
+     */
+    public function __invoke(Call $call, array $options)
+    {
+        $next = $this->nextHandler;
+
+        if (empty($this->autoPopulationSettings)) {
+            return $next($call, $options);
+        }
+
+        $request = $call->getMessage();
+        foreach ($this->autoPopulationSettings as $fieldName => $valueType) {
+            $getFieldName = 'get' . ucwords($fieldName);
+            // Populate the field if it's not already set by user
+            if (empty($request->$getFieldName())) {
+                $setFieldName = 'set' . ucwords($fieldName);
+                switch ($valueType) {
+                    case \Google\Api\FieldInfo\Format::UUID4:
+                        $request->$setFieldName(Uuid::uuid4()->toString());
+                        break;
+                    default:
+                        // Do nothing
```

I think we should throw an exception for invalid values, wdyt?

---

This only case where I see this happening is when all the the below happens at the same time:
1) Some new enum get's added to the `google.api.field_info.Format`
2) This library's minimum version of common-protos-php isn't updated.
3) The new field_info is used for autopopulation

So we would always want to catch this error!!!

---

Added the logic